Expose bot names to localization

Mailaender opened the pull request:
as not all of them are proper names.

PunkPun commented on OpenRA.Mods.Common/Traits/Player/ModularBot.cs:
```diff
@@ -25,15 +25,16 @@ public sealed class ModularBotInfo : TraitInfo, IBotInfo
 		string IBotInfo.Type => Type;
 
-		string IBotInfo.Name => Name;
+		string IBotInfo.Name => TranslationProvider.GetString(Name);
```
translating here will very likely cause issues as this name is used not only for UI

RoosterDragon replied:
Not sure I see the issue here. This maintains the previous behaviour where this field returned the display name.

And this doesn't seem like a performance issue, none of the callers looks like they'd get called frequently.

PunkPun replied:
it's more that there can be issues with syncing with other players as they might be using a different language

pchote replied:
Are you sure that you (or perhaps the code) aren't getting this mixed up with `IBotInfo.Type`? `Name` is supposed to only be used for UI, and `Type` for matching to internal things.

RoosterDragon replied:
I think Pun is right:

`ResolvePlayerName` uses the name for checking if bots are of the same type. That sounds like it would be safer if it used the type.

`SanitizedPlayerName` stops people using a bot name, if the bot names can differ in language that might need changing somehow.

pchote replied:
`SanitizedPlayerName` is an interesting case, we probably want some new plumbing to query a dictionary of language id -> string for a given translation id.

The alternative would be to remove the bot name check completely...

RoosterDragon replied:
Dropping the bot name check seems reasonable.

If we did want to have some sort of safety here, adding a bot icon or some other UI affordance would allow distinguishing between human and AI players. Such an affordance would also be more helpful than the current don't-reuse-name check, e.g. if a human player uses a visually similar yet slightly different name to one of the bots.

pchote replied:
That makes sense - we already do this for registered vs anonymous players, so that indicator could grow a new 🤖 state.

PunkPun replied:
but it still seems wrong to have the names of the bots be determined by the language the admin runs. It could have font issues as well

pchote replied:
Indeed, bot names should be resolved client-side. This should be simple enough, since the bot type is specified on the client info.

RoosterDragon commented on OpenRA.Mods.Common/Traits/Player/ModularBot.cs:
```diff
@@ -25,15 +25,16 @@ public sealed class ModularBotInfo : TraitInfo, IBotInfo
+		[TranslationReference]
 		[Desc("Human-readable name this bot uses.")]
-		public readonly string Name = "Unnamed Bot";
+		public readonly string Name = null;
```
Worth applying this change in `DummyBot` as well? Which appears to be used in Fort Lonestar.

Mailaender replied:
There is a limitation of the map rules translation not showing in the lobby in this case, so I reverted that change.

pchote replied:
This sounds like another case that will need additional plumbing - since there are other cases (e.g. map option labels) that will need this.

The cleanest solution would be to extend `MapPreview` and the resource center API to load the map specified translations.

Mailaender replied:
This is fixed by translating deeper down the chain.

RoosterDragon replied:
We still need the `[TranslationReference]` in `DummyBotInfo` I believe?

Mailaender replied:
It still does not work. I won't be able to solve it in this pull request.

pchote replied:
The `MapPreview` changes will also be needed for any custom map defined bots.

pchote commented on mods/cnc/languages/rules/en.ftl:
```diff
@@ -559,3 +559,13 @@ actor-truck =
    .name = Supply Truck
+
+## ai.yaml
+cabal =
```
Would it make sense to give these a `bot-` prefix, like actors have `actor-` above?

Mailaender committed:
Don't disallow players chosing bot names.

Mailaender committed:
Add a UI indicator for bot players.

Mailaender commented:
![grafik](https://github.com/user-attachments/assets/d727a181-1e44-42a3-81f8-f583104436c8)

RoosterDragon commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs:
```diff
@@ -352,15 +352,17 @@ public static void SetupLatencyWidget(Widget parent, Session.Client c, OrderMana
 			if (profile != null)
 			{
-				var imageName = (c != null && c.IsAdmin ? "admin-" : "player-")
-					+ (c.Fingerprint != null ? "registered" : "anonymous");
```
This method can't decide on whether `c` might be null or not, sometimes it checks but other times it doesn't.

A preexisting issue, but might be nice to fix that whilst you're in here. At a cursory glance it appears all callers provide a non null client so I think removing the null checks is safe.

Mailaender replied:
Yes, indeed the whole thing wasn't that readable.

RoosterDragon commented on mods/cnc/languages/chrome/en.ftl:
```diff
@@ -696,5 +696,7 @@ button-settings-panel-reset = Reset
 label-latency-tooltip-prefix = Latency:
 label-anonymous-player-tooltip-name = Anonymous Player
+label-bot-player-tooltip-name = Bot
+label-bot-controlled-by-tooltip = Bot controlled by { $name }
```
"Controlled" sounds like the human player can give them orders in game to me, maybe a different word like "Bot managed by { $name }"? Or maybe even just "Bot" or "AI Player".

Or I could be overthinking it.

Mailaender replied:
I had similar thoughts, but was uncreative.

PunkPun reviewed (commented):
It looks like we are still translating on bot owner side. So the names will be localised incorrectly for other players and replays

<img width="322" alt="Screenshot 2024-07-26 at 21 48 00" src="https://github.com/user-attachments/assets/703df264-b9f6-45cb-84c7-3ee751fd0998">

this font also seems humongous

PunkPun reviewed (commented):
<img width="364" alt="Screenshot 2024-07-26 at 21 52 08" src="https://github.com/user-attachments/assets/9da82085-5ad4-4a29-b359-f9127ee89e8d">

Mailaender commented:
You are right. I moved the translation into UI code now.

PunkPun commented on OpenRA.Game/Player.cs:
```diff
@@ -138,7 +138,8 @@ public static string ResolvePlayerName(Session.Client client, IEnumerable<Sessio
 				var botsOfSameType = clients.Where(c => c.Bot == client.Bot).ToArray();
-				return botsOfSameType.Length == 1 ? botInfo.Name : $"{botInfo.Name} {botsOfSameType.IndexOf(client) + 1}";
+				var botName = TranslationProvider.GetString(botInfo.Name);
+				return botsOfSameType.Length == 1 ? botName : $"{botName} {botsOfSameType.IndexOf(client) + 1}";
```
The issue is translating here. The results of this function are used in networking code

pchote replied:
We should also be passing the numeric suffix as an argument to the translation instead of appending it manually.

PunkPun reviewed (commented):
this hasn't fixed it

<img width="378" alt="Screenshot 2024-07-26 at 23 20 02" src="https://github.com/user-attachments/assets/886506be-3bee-4009-843c-0eb3278be6da">

Mailaender commented:
Updated to keep the replays also client side translated.
![grafik](https://github.com/user-attachments/assets/f57802ce-6aa5-4da6-b73f-2e53f848f3f4)

Also added localization for player name suffixes.

PunkPun commented:
the translated name still appears in replays at the very bottom. I'm also afraid of it appearing in sync checks

PunkPun reviewed (commented):
the translated name still appears in replays at the very bottom. I'm also afraid of it appearing in sync checks

PunkPun commented:
maybe also try watching the `TranslationProvider` during a multiplayer match to check wether it receives any requests to translate already translated AI names

Mailaender committed:
Expose player names to localization.

PunkPun commented on OpenRA.Game/GameInformation.cs:
```diff
@@ -143,6 +147,19 @@ public Player GetPlayer(OpenRA.Player runtimePlayer)
+				var botsOfSameType = Players.Where(p => p.BotType == player.BotType).ToArray();
+				return TranslationProvider.GetString(EnumeratedBotName,
+					Translation.Arguments("name", TranslationProvider.GetString(player.Name),
+						"number", botsOfSameType.IndexOf(player) + 1));
```
```suggestion
				var number = Players.Where(p => p.BotType == player.BotType).ToList().IndexOf(player) + 1;
				return TranslationProvider.GetString(EnumeratedBotName,
					Translation.Arguments(
						"name", TranslationProvider.GetString(player.Name),
						"number", number));
			}
```

I think this would be much more readable formatting

PunkPun commented on OpenRA.Game/Player.cs:
```diff
@@ -224,7 +218,21 @@ public Player(World world, Session.Client client, PlayerReference pr, MersenneTw
+						"number", botsOfSameType.IndexOf(this) + 1));
+			}
+
+			return PlayerName;
```
I think `ResolvedPlayerName` should be cached. It doesn't change from game start and is called often

PunkPun commented on mods/common/languages/en.ftl:
```diff
@@ -860,3 +860,21 @@ notification-desync-compare-logs = Out of sync in frame { $frame }.
+    } { $client-state ->
+        [disconnected] (Gone)
+       *[other] {""}
+    }
```
this conflicts with https://github.com/projectfluent/fluent/wiki/Good-Practices-for-Developers

this logic should be inside c# and not ftl files

Mailaender replied:
I thought the word order might be different in other languages.

PunkPun replied:
I meant `won` or `disconnected` shouldn't be arguments

PunkPun reviewed (commented):
otherwise LGTM

Mailaender commented:
Also contains new artwork by @Dzierzan
![grafik](https://github.com/user-attachments/assets/090e827c-c15d-442b-826c-2b85f0cc9338)

pchote commented:
Were the artwork changes made to the artsrc svg, or directly in the exported pngs?

PunkPun commented on OpenRA.Mods.Common/Widgets/WidgetUtils.cs:
```diff
@@ -324,19 +333,26 @@ public static void BindPlayerNameAndStatus(LabelWidget label, Player p)
+					text += $" ({TranslationProvider.GetString(Lost)})";
+
+				if (client.State == Session.ClientState.Disconnected)
+					text += $" ({TranslationProvider.GetString(Gone)})";
```
the original behaviour only uses 1 suffix, and this implementation now allows multiple. If we keep the original behaviour then we can keep formatting on the tfl side

Mailaender commented:
> Were the artwork changes made to the artsrc svg, or directly in the exported pngs?

https://github.com/OpenRA/ArtSrc/pull/23

PunkPun commented:
https://github.com/OpenRA/ArtSrc/pull/23 was updated

PunkPun commented:
[chnagelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/793559730a9de02352cd88be2f811af6a84f0626...90a89f2266e0b6578fb60d193fd212a6571e57ca)
